### ISSUE TITLE
Lint path or import used in macro

### ISSUE THREAD

**Issue by i509VCB** [rust-lang/rust-clippy#9104]
### What it does

Lint any instances of paths to items or use statements which do not use `$crate` or global paths (such as `::std::vec::Vec`)

### Lint Name

macro_use_path

### Category

correctness

### Advantage

- Since imports and paths in macros are unhygienic, you should use `$crate` or global paths.
- This will make macros used by other crates less likely to fail to compile.

### Drawbacks

- Internal macros may wish to avoid this requirement

### Example

```rust
pub trait Delegate {}

#[macro_export]
macro_rules! impl_delegate {
    ($ty: ty) => {
        impl Delegate for $ty {}
    }
}
```

Could be written as:

```rust
pub trait Delegate {}

#[macro_export]
macro_rules! impl_delegate {
    impl $crate::Delegate for $ty {}
}
```

**Comment by i509VCB** [comment]
Discussion on zulip about this lint is going to be a challenge given the macro is not enough information (you need to look at the expanded macro to determine the AST.

**Comment by BenjaminBrienen** [comment]
Would have prevented this issue: https://github.com/bevyengine/bevy/issues/15615

**Comment by BenjaminBrienen** [comment]
This would help avoid the complexity: https://github.com/scrabsha/expandable

**Comment by scrabsha** [comment]
> This would help avoid the complexity: https://github.com/scrabsha/expandable
> @scrabsha 

Thanks for the ping! This is something I plan to add for the 1.0 release of expandable, which is far from ready.

Also keep in mind that the algorithms implemented in `expandable` can't be used in "official tools" because they are slightly too opinionated.

**Comment by kpreid** [comment]
> this lint is going to be a challenge given the macro is not enough information (you need to look at the expanded macro to determine the AST.

An inelegant but effective mechanism to achieve this would be to lint based on the expansion at a use site, but **only if** the lint is active in the crate defining the macro — in particular, to respect `--cap-lints` on the defining crate. This way, the lint only works if the macro has at least one use (hopefully including tests or examples present in the defining package), but it won’t bother users that shouldn’t be bothered.
